Initial urdf parser plugin that parses sdformat

sloretz opened the pull request:
This is ready for review.

@scpeters @iantheengineer FYI

requires ros2/urdf#13 (This is merged :tada:)
~~requires ros/resource_retriever#50~~ (no longer uses this)

sloretz committed:
Initial urdf parser plugin that parses sdformat

Lots of TODOs

Signed-off-by: Shane Loretz<sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Visual and Collision with Box geometry

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Sphere geometry

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Cylinder geometry

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Mesh support

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-07-16/15468/1

scpeters commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,466 @@
+    }
+  }
+
+  // sdf canonical link -> urdf root link
```
this may be true for specific models but is not true in general. I think we could add some helper APIs to libsdformat to aid in determining:

* if the directed graph in an SDFormat file is also a directed tree
* and if so, which is the root link of that directed tree

we are brainstorming what those APIs could look like in https://github.com/osrf/sdformat/issues/273#issuecomment-663292059

for now, this assumption is fine

EricCousineau-TRI replied:
Perhaps this can be worded as a `TODO` pointing to that issue? e.g.
```
TODO(name) Possibly relax this assumption given the outcome of osrf/sdformat#273.
```

sloretz replied:
Reworded assumption + added TODO c84a46db49923115a56319d0ef0f07a6136b97d1

EricCousineau-TRI commented on package.xml:
```diff
@@ -0,0 +1,36 @@
+  <name>sdformat_urdf</name>
+  <version>0.1.0</version>
+  <description>
+    URDF plugin to parse SDFormat XML into URDF C++ structures.
```
Is it fair to restate `URDF C++ structures` as `URDF C++ DOM objects`?

sloretz replied:
`structures` -> `DOM objects` in aff9def

EricCousineau-TRI commented on package.xml:
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>sdformat_urdf</name>
```
(Feel free to dismiss) As the peanut gallery, can't say I love this name... Some other candidates:

- `sdformat_urdf_converter`
- `sdformat_to_urdfdom`
- `sdformat_urdfdom`
- `sdformat_urdfdom_plugin`

clalancette replied:
To add my own thoughts to this bikeshed:

I don't think we should use the name "urdfdom" in it, as to me, that implies converting to a URDF XML format (which is not what we are up to here).  I'm fine with:

* `sdformat_urdf`
* `sdformat_urdf_plugin`
* `sdformat_to_urdf`

scpeters replied:
I like `sdformat_to_urdf`

sloretz replied:
It doesn't seem like any of the names are obviously winners. At least `sdformat_urdf` matches `collada_urdf`, so I'm inclined to leave it as is for consistency.

EricCousineau-TRI commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,466 @@
+    case sdf::JointType::PRISMATIC:
+      urdf_joint->type = urdf::Joint::PRISMATIC;
+      break;
+    case sdf::JointType::INVALID:
```
Dumb question: Why even specify these cases as fall-through when they all feed into the `default` case?

It seems more harmful to have them here, as someone may naively fill in an intermediate case (e.g. `GEARBOX`) and mess up the fall-through?

sloretz replied:
It's to show they're known and intentionally handled by the default case instead of accidentally handled by the default case.

I hope a contribution adding code in the middle would be caught in code review, and maybe caught with unit testing.

sloretz replied:
Changed fall through comments to make it easier to see the cases are all fall through to `default`

ecb6e32e8ede94cf00ac7051510312acc1f33af5

EricCousineau-TRI commented on include/sdformat_urdf/sdformat_urdf.hpp:
```diff
@@ -0,0 +1,33 @@
+namespace sdformat_urdf {
+/// \brief Parse an SDFormat XML string and return URDF C++ structures
+urdf::ModelInterfaceSharedPtr
+parse(const std::string & data, sdf::Errors & errors);
```
Dumb question: Do these need any visibility macros, given that these look like public API?

sloretz replied:
Added visibility macros in 3c0a28f

EricCousineau-TRI commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,466 @@
+
+  // Start with root link (sdf canonical frame) and recurse this function:
+  //  
+  // TODO what about coordinate frames?
```
Er, I'm not sure if I understand this TODO. Why are coordinate frames important? I think you just resolve the pose and you're good (which is what you're doing below)?

sloretz replied:
I forgot to delete the TODO after adding the code below

sloretz replied:
Deleted TODO in 4f0ef012274e6f79d8c38010966a096feb9c26a7

sloretz committed:
remove TODO comment about resolving poses

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Let downstream worry about resolving mesh URIs

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Reword assumption about kinematic root link

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Make fall through comments easier to see

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Create urdf_mesh variable

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Handle inertia

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Mesh scale

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Visibility control header

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,462 @@
+    case sdf::JointType::PRISMATIC:
+      urdf_joint->type = urdf::Joint::PRISMATIC;
+      break;
+    case sdf::JointType::INVALID:     // Unsupported: fall through to default
+    case sdf::JointType::BALL:        //  |
+    case sdf::JointType::GEARBOX:     //  |
+    case sdf::JointType::REVOLUTE2:   //  |
+    case sdf::JointType::SCREW:       //  |
+    case sdf::JointType::UNIVERSAL:   //  V
```
I love the ASCII art :).

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,462 @@
+}
+
+urdf::Pose
+sdformat_urdf::convert_pose(const ignition::math::Pose3d & sdf_pose)
```
Nit: `#include <ignition/math/Pose3.hh>`

sloretz replied:
Included header in ab6637bcb03d00f08facd94da7956ffe25cb6bf4

clalancette commented on src/sdformat_urdf_plugin.cpp:
```diff
@@ -0,0 +1,58 @@
+
+namespace sdformat_urdf {
+
+class SDFormatURDFParser : public urdf::URDFParser
+{
+public:
+
+  SDFormatURDFParser() = default;
+  virtual ~SDFormatURDFParser() = default;
```
This doesn't seem like a class that would be subclassed, so I'll suggest marking the class `final` and removing the virtual destructor.

sloretz replied:
Use `final` and remove `virtual` from destructor in 21adf189d3d02e72014e5a642a15123c7ca5cf32

clalancette commented on src/sdformat_urdf_plugin.cpp:
```diff
@@ -0,0 +1,58 @@
+  size_t
+  might_handle(const std::string & data) override
+  {
+    return data.find("<sdf");
```
As with the URDF plugin, I think this is a bit too simplistic.  I'd recommend doing the same thing as we did over there, using tinyxml2 to do the parsing to check for the root element.

sloretz replied:
Use tinyxml2 to filter valid XML documents that are not SDFormat in 7a5ad72442456120647b863d9522085c4c5284ff

clalancette reviewed (commented):
A few mostly minor things in here.

sloretz committed:
Remove resource_retriever dependency

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Support joint axis expressed in other frames

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Include ignition/math/Pose3.hh

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use final on plugin class

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use TinyXML2 to filter non-sdf XML documents

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Apache 2.0 License

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Satisfy uncrustify

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Satisfy cpplint

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add linter tests

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

clalancette commented on src/sdformat_urdf_plugin.cpp:
```diff
@@ -0,0 +1,76 @@
+  {
+    urdf::ModelInterfaceSharedPtr urdf_cpp;
+    sdf::Errors errors;
+    urdf_cpp = sdformat_urdf::parse(data, errors);
```
Nit: you could save a line of code here:

```suggestion
    sdf::Errors errors;
    urdf::ModelInterfaceSharedPtr urdf_cpp = sdformat_urdf::parse(data, errors);
```

sloretz replied:
Saved the line of code in c9ca571

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+  return nullptr;
+}
+
+
```
Nit: extra blank line.

sloretz replied:
Removed extra blank lines in 135cb90

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+      return nullptr;
+    }
+
+    auto pair = urdf_model->links_.emplace(sdf_link->Name(), convert_link(*sdf_link, errors));
```
Nit: As usual, I don't like the use of auto.  But feel free to ignore if you think it is clear enough.

(same goes a couple more times below)

sloretz replied:
I'll ignore this one. I think where `auto` is being used the types are still easy to see (and/or unimportant and very verbose in the case of pairs and iterators).

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+
+  const ignition::math::Inertiald sdf_inertia = sdf_link.Inertial();
+  urdf_link->inertial = std::make_shared<urdf::Inertial>();
+  if (!urdf_link->inertial) {
```
I'm not sure that this check is useful here.  From the documentation, I think that `std::make_shared` will throw an exception on memory allocation failure: https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared

sloretz replied:
Removed impossible check in 0dd8103

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+    if (!sdf_visual) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get visual on lilnk [" + sdf_link.Name() + "]");
```
```suggestion
        "Failed to get visual on link [" + sdf_link.Name() + "]");
```

sloretz replied:
lilnk -> link f7fcfe5

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+      errors.insert(errors.end(), pose_errors.begin(), pose_errors.end());
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get transfrom from visual [" + sdf_visual->Name() +
```
```suggestion
        "Failed to get transform from visual [" + sdf_visual->Name() +
```

sloretz replied:
Transfrom -> transform in 4d9416aba41190f748afca554a736a85af998d37

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+    const sdf::Material * sdf_material = sdf_visual->Material();
+    if (sdf_material) {
+      // TODO(sloretz) textures
+      // TODO(sloretz) error if any file names we can't resolve are given
```
I'm fine with punting these TODOs with later, but I will suggest opening issues about them to give them more visibility.

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+
+  urdf_joint->name = sdf_joint.Name();
+
+  size_t num_axes = 0;
```
This is a bit of a nit, but this is really a boolean: has_one_axis.  It might be a bit simpler down below to switch to that.

sloretz replied:
Got rid of variable in ce65068

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "Plane geometry cannot be converted to urdf C++ structures");
+    return nullptr;
```
You might consider putting this error case at the end, and then falling through to the `return nullptr`.  That is, the end here would look like:

```
} else if (sdf_geometry.PlaneShape()) {
  errors.emplace_back(
    sdf::ErrorCode::STRING_READ,
    "Plane geometry cannot be converted to urdf C++ structures");
} else {
  errors.emplace_back(
    sdf::ErrorCode::STRING_READ,
    "Unknown geometry shape");
}

return nullptr;
```

(I'm honestly not sure if that is easier to read; I just found it weird that there is an error case in the middle of the `if` structure followed by a successful case)

sloretz replied:
Moved PlaneShape check to bottom in  in 44763d4

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+
+    // Check if there is a kinematic loop
+    if (visited_links_.end() !=
+      std::find(visited_links_.begin(), visited_links_.end(), sdf_parent_link))
```
I'm somewhat concerned that on large models, this is going to be expensive.

The good news is that I think it is easy to fix.  That is, if you switched `visited_links` to a `std::set` or a `std::unordered_map`, you could still answer the same question of "Did we visit this link already?".  `std::set` fits a bit more naturally, but `std::unordered_map` has constant-time lookup, so I'll leave it to you what you want to use here.

(don't forget to `#include` the appropriate header at the top of the file)

(also, as a nit, function-local variables don't get an `_` in ROS style)

sloretz replied:
I had to refactor this code to fix a bug in 1b36817f15d02f66e22223a4a47ac55816c87897 - there's no longer an `std::find()`. Finding kinematic loops is done by checking if a link is a child of multiple joints instead.

clalancette commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,489 @@
+          errors.insert(errors.end(), pose_errors.begin(), pose_errors.end());
+          errors.emplace_back(
+            sdf::ErrorCode::STRING_READ,
+            "Failed to get transfrom from joint [" + sdf_joint->Name() +
```
```suggestion
            "Failed to get transform from joint [" + sdf_joint->Name() +
```

sloretz replied:
transfrom -> transform in 4d9416aba41190f748afca554a736a85af998d37

clalancette reviewed (dismissed):
I've done another review here and come up with some fixes.  Most of them are nits, but there are a couple of things that need to be looked at.

sloretz committed:
Add test for pose_link case

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Actually add test files

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add  SDF link pose to URDF link members

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Pose link test actually checks pose

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add SDF visual pose test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add SDF collision pose test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add SDF inertial pose test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add test for model with pose, plus refactor test tools

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Replace todo with explanation

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add pose_link_all test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Assert model name is correct

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add pose_link_in_frame test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add pose_collision_in_frame test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add pose_visual_in_frame test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add skipped pose_inertial_in_frame test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Depend on sdformat

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix link poses of child links

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add test for model with joint pose

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unused variable

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add test for model with joint pose in frame

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Rename variables for consistent ordering

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix fame names in calculating joint pose

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
@azeey @scpeters I've been having trouble writing the expectations in the `<pose>` tests, but I think I'm starting to understand. Does anything below sound incorrect or unreasonable? Any thoughts?

I think my confusion comes from SDFormat link poses vs URDF Link poses. SDFormat links can have poses, while URDF links don't have poses. The URDF link frame is really a joint frame - the pose of a urdf link is the pose of the urdf joint having that urdf link as a child. The root urdf link has no parent joint so it has nothing for the link to be offset from. This means link frames visualized in RViz are the same as joint frames in Gazebo, while Link frames in Gazebo (with the exception of the canonical frame) don't exist in the URDF structure at all.

Since in URDF the root link can't be offset from anything, it seems like the `//link/pose` of the root (sdformat_urdf assumes canonical is the root) won't make it into the URDF c++ structures. It might help position links and joints in the SDFormat XML, but URDF C++ structures only have the relative positions of those links and joints from the root.

It seems like there is more potential for confusion between SDFormat XML link frames and URDF/tf link frames. In URDF the link frames have the link's name but are really the locations of joints. Given `link1 -> joint1 -> link2` the PR right now creates the tf tree `link1 -> link2` where URDF `link2` is the location of SDFormat `joint1`. Someone who hand wrote SDFormat XML might be be surprised when they ask for the tf transform of `link1` to `link2` and get the value of sdf `link1` to sdf `joint1`.

Maybe this could be fixed by adding dummy links? That is the tf tree for the sdf `link1, link2, link3` with `joint1 = link1 to link2` and `joint2 = link2 to link3` would have `[link frames]` and `-joints->` like:

```
[link1]-revolute->[joint1]-fixed->[link2]
                \
                 -revolute->[joint2]-fixed>[link3]
```
or
```
[link1]-revolute->[joint1]-fixed->[link2]-revolute->[joint2]-fixed>[link3]
```

Then the transform from urdf/tf `link1` to urdf/tf `link2` represents the transform that would be seen from sdf `link1` to sdf `link2`. A downside is either urdf `joint1`/`joint2` or urdf `link2`/`link3` would have to have no visuals, inertia, or collisions. It's also not clear to me what to call the link frames for the joints.

azeey commented:
I don't see anything incorrect in what you wrote. In URDF, the joint frame and the child link frame are colocated, whereas in SDFormat, the joint frame can be offset from the child link. This offset is given by `//joint/pose`. FYI, this is described in a little more detail [here](http://sdformat.org/tutorials?tut=spec_model_kinematics&cat=specification&#example-joint-revolute)

There's some work in an sdformat branch [here](https://github.com/osrf/sdformat/compare/sdformat4_4.4.0...sdf_to_urdf), which I later updated to libsdformat8 [here](https://github.com/osrf/sdformat/compare/sdformat8_8.4.0...sdf_to_urdf_sdf8) that converts SDFormat to URDF. Sorry, I should have mentioned it sooner. In that code, the link's pose is used to offset the inertial, visual and collision elements in URDF. This works fine for visualization, but like you observed, the transform between two links will not always be the same in SDFormat and URDF. 

Adding dummy links could solve the problem with the caveat that converting that URDF back to SDFormat will not generate the original SDFormat, but I think that's okay.

scpeters commented:
> I don't see anything incorrect in what you wrote. In URDF, the joint frame and the child link frame are colocated, whereas in SDFormat, the joint frame can be offset from the child link. This offset is given by `//joint/pose`. FYI, this is described in a little more detail [here](http://sdformat.org/tutorials?tut=spec_model_kinematics&cat=specification&#example-joint-revolute)

I agree

> There's some work in an sdformat branch [here](https://github.com/osrf/sdformat/compare/sdformat4_4.4.0...sdf_to_urdf), which I later updated to libsdformat8 [here](https://github.com/osrf/sdformat/compare/sdformat8_8.4.0...sdf_to_urdf_sdf8) that converts SDFormat to URDF. Sorry, I should have mentioned it sooner. In that code, the link's pose is used to offset the inertial, visual and collision elements in URDF. This works fine for visualization, but like you observed, the transform between two links will not always be the same in SDFormat and URDF.

> Adding dummy links could solve the problem with the caveat that converting that URDF back to SDFormat will not generate the original SDFormat, but I think that's okay.

I think losing the link frame location is just a limitation of urdfdom. It will be implicitly included in the pose of collisions, inertias, and visuals. I don't think dummy links are strictly necessary

sloretz committed:
URDF link frames are coincident with joint frames

And the root link pose has no effect on the URDF output except that it
changes the relative locations of other links with the root link.
The SDF link pose cannot be expressed in URDF C++ structures without
adding dummy links, but this does not do that.

Signed-off-by: Shane Loretz<sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add pose_chain test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Clearer description of pose_collision expectations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Clearer description of pose_collision_in_frame expectations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Clearer pose_inertial* expectations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unnecessary ASSERT

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Initializer lists

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unnecessary ASSERT

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Initializer lists

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix pose_link* test expectations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Clearer pose_visual* test expecations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Shorter file name

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Shorter file name

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_loop test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_four_bar test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Alphabetize

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> I think losing the link frame location is just a limitation of urdfdom. It will be implicitly included in the pose of collisions, inertias, and visuals. I don't think dummy links are strictly necessary

This seems reasonable to me. It looks like this is the most straightforward path in code too. I've updated some test expectations that were wrong before.

sloretz committed:
Refactor tree traversal code

Fixes bug were tree's with non-canonical roots were not caught during
conversion
Shrink list of joints to check after checking them
Use visited_links only to catch multiple roots and algorithm errors
Get rid of trailing_

Signed-off-by: Shane Loretz<sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_tree_non_canonical_root test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Transfrom -> transform

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Different error when XML file has no model

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
urdf child link keeps weak reference to parent link

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_tree test and EXPECT_NAMES util

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_chain test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add graph_chain_non_canonical_root test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_plane test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_box test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_cylinder test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_sphere test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_heightmap test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_collada test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_obj test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_stl test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add geometry_mesh_scaled test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add tests for zero and two models

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_ball test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Get rid of num_axes variable

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add damping and friction info to joints

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_continuous test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

azeey commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,569 @@
+{
+  auto sdf_dom = std::make_shared<sdf::Root>();
+  errors = sdf_dom->LoadSdfString(data);
+  if (sdf_dom) {
```
`sdf_dom` is going to be a valid pointer regardless of what happens in `LoadSdfString`. Maybe you could check if `errors` is empty?

sloretz replied:
Checking errors instead of pointer in ce0812d

sloretz committed:
Add joint_fixed test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

azeey commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,574 @@
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get sdf joint");
+      return nullptr;
```
Probably redundant since the this is checked in the previous joint iteration.

sloretz replied:
It looks like this comment was on an earlier version of the file. It doesn't appear to be a redundant anymore, but I don't know which commit fixed it.

azeey replied:
Sorry, I meant the for loop in https://github.com/ros/sdformat_urdf/blob/aff9defb8b4d4c5ca7740cbab05020b136b8a789/sdformat_urdf/src/sdformat_urdf.cpp#L166-L171. Feel free to ignore if you think the extra check is necessary.

sloretz replied:
Oh that one! I'll leave the extra check in now. It seems inexpensive, and it protects the code if something above gets moved around.

sloretz committed:
Add joint limits info

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_prismatic test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

azeey commented on src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,574 @@
+        urdf_parent_link->child_joints.push_back(urdf_joint);
+
+        // SDFormat joint pose is relative to the child sdformat link
+        // URDF joint pose is relative to the parent urdf link, which is the the frame of the
```
nit: which is the **the** frame ...

sloretz replied:
the the -> the 6552021

sloretz committed:
Add axis expectations

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute2 test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_screw test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_universal test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add link_inertia test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute_axis test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Warn on non-default values of unsupported <joint> tags

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add link_multiple_collisions test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add link_multiple_visuals test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute_axis_in_frame test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute_default_limits test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Detect kinematic loops involving redundant joints

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_revolute_two_joints_two_links test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Warn if <sensor> or <light> are used on <link>

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add link_sensor_imu test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add link_light_point test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Sensors -> lights

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Move package into folder in repo

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add sdformat_test_files package to repo

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
lilnk -> link

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Whitespace

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Save a line

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Move PlaneShape check lower

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove impossible condition after make_shared

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

clalancette reviewed (commented):
The structure of the code here looks good to me.  I'm not going to approve, since I think you really need someone who knows SDF (@scpeters or @azeey) to go over the code more closely.

Thanks for fixing the stuff from my previous review!

sloretz committed:
Make names shorter

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove accidentally comitted file

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add material_blinn_phong test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Warn if unsupported <material><lighting> feature is used

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add skipped material_dynamic_lights test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unfinished material models and add TODO for them

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Expect no Algorithm errors

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add joint_gearbox model and test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Update descriptions of all models

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add sdformat_test_files README

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
Checking with ROS 2 CI (testing only sdformat_urdf) **Edit: failing because CI jobs don't have `libsdformat9-dev` installed :(**
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12725)](http://ci.ros2.org/job/ci_linux/12725/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7686)](http://ci.ros2.org/job/ci_linux-aarch64/7686/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10439)](http://ci.ros2.org/job/ci_osx/10439/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12672)](http://ci.ros2.org/job/ci_windows/12672/)

sloretz committed:
Add README for sdformat_urdf

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Misc tiny edits

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Eliminate elements

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add repo level readme

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove .

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Trick to use sdformat9 from source

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove condition hack

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
Trying CI again - this time using a merged workspace and a repos file that includes `osrf/sdformat` on the `sdf9` branch - **Edit: failing because build args didn't include sdformat9**

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12727)](http://ci.ros2.org/job/ci_linux/12727/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7688)](http://ci.ros2.org/job/ci_linux-aarch64/7688/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10441)](http://ci.ros2.org/job/ci_osx/10441/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12674)](http://ci.ros2.org/job/ci_windows/12674/)

sloretz commented:
CI Again - merged workspace trick with repos file including `osrf/sdformat` on `sdf9` branch with build args building sdformat9 and sdformat_urdf - hopefully `sdformat9` finishes building before `sdformat_urdf` starts. This would be a lot easier if the rosdep key for sdformat and the CMake `project()` name for sdformat were the same. **Edit: Failed because missing ignition-tools**

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12740)](http://ci.ros2.org/job/ci_linux/12740/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7700)](http://ci.ros2.org/job/ci_linux-aarch64/7700/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10453)](http://ci.ros2.org/job/ci_osx/10453/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12686)](http://ci.ros2.org/job/ci_windows/12686/)

scpeters commented:
> CI Again - merged workspace trick with repos file including `osrf/sdformat` on `sdf9` branch with build args building sdformat9 and sdformat_urdf - hopefully `sdformat9` finishes building before `sdformat_urdf` starts. This would be a lot easier if the rosdep key for sdformat and the CMake `project()` name for sdformat were the same.
> 
> * Linux [![Build Status](https://camo.githubusercontent.com/b8ff22022df82321889ee41513fe0bc93fe8d0e2/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e7578266275696c643d3132373430)](http://ci.ros2.org/job/ci_linux/12740/)
> * Linux-aarch64 [![Build Status](https://camo.githubusercontent.com/8ab9deb330bb91c65202da829f5b8555f40f12d5/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e75782d61617263683634266275696c643d37373030)](http://ci.ros2.org/job/ci_linux-aarch64/7700/)
> * macOS [![Build Status](https://camo.githubusercontent.com/0416c607cd2398a6b93b98906909b42ec415480c/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6f7378266275696c643d3130343533)](http://ci.ros2.org/job/ci_osx/10453/)
> * Windows [![Build Status](https://camo.githubusercontent.com/2098ee865ea75fca7bfc6c5f4aecd9d0c08bbf4e/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3132363836)](http://ci.ros2.org/job/ci_windows/12686/)

can you add a `colcon.pkg` file to this project with the proper cmake project name for sdformat9?

sloretz committed:
Add colcon.pkg for building sdformat from source

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> can you add a colcon.pkg file to this project with the proper cmake project name for sdformat9?

Added in c3ca66a

sloretz commented:
CI again - now with `colcon.pkg` plus repo contains ignition-tools1, ignition-math6, and ignition-cmake2 **edit: failed because wrong branch name in repos file**

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12743)](http://ci.ros2.org/job/ci_linux/12743/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7704)](http://ci.ros2.org/job/ci_linux-aarch64/7704/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10456)](http://ci.ros2.org/job/ci_osx/10456/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12689)](http://ci.ros2.org/job/ci_windows/12689/)


**Edit attempt 2 with fixed repos file**  **Edit again: Failing because ignition-tools requires ruby**

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12745)](http://ci.ros2.org/job/ci_linux/12745/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7706)](http://ci.ros2.org/job/ci_linux-aarch64/7706/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10458)](http://ci.ros2.org/job/ci_osx/10458/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12691)](http://ci.ros2.org/job/ci_windows/12691/)

azeey commented on sdformat_test_files/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+  "${CMAKE_CURRENT_BINARY_DIR}/sdformat_test_files/sdformat_test_filesConfigVersion.cmake"
+  VERSION "${sdformat_test_files_VERSION}"
+  COMPATIBILITY SameMajorVersion
+  ARCH_INDEPENDENT)
```
`ARCH_INDEPENDENT` is only available in cmake >= 3.14. We can change `cmake_minimum_required` version, but that would mean this package won't build on Ubuntu Bionic. Is the flag absolutely necessary?

sloretz replied:
It's not absolutely necessary. Only uses ARCH_INDEPENDENT if CMake version supports it in a533c05

azeey commented on sdformat_urdf/CMakeLists.txt:
```diff
@@ -0,0 +1,175 @@
+ament_export_dependencies(urdfdom_headers)
+ament_export_dependencies(sdformat9)
+
+install(TARGETS sdformat_urdf sdformat_urdf_plugin
+  ARCHIVE DESTINATION lib
+  LIBRARY DESTINATION lib
+  RUNTIME DESTINATION bin
+)
+
+install(
+  DIRECTORY include/
+  DESTINATION include
+)
+
+install(TARGETS sdformat_urdf EXPORT sdformat_urdf-export)
+ament_export_targets(sdformat_urdf-export)
+
```
The command `install(TARGETS sdformat_urdf EXPORT sdformat_urdf-export)` fails on Bionic with the message `  install TARGETS given no LIBRARY DESTINATION for shared library target "sdformat_urdf".` Here's my suggestion based on the ament [documentation](https://index.ros.org/doc/ros2/Tutorials/Ament-CMake-Documentation/#building-a-library):

```suggestion
install(TARGETS sdformat_urdf sdformat_urdf_plugin
  EXPORT sdformat_urdf-export
  ARCHIVE DESTINATION lib
  LIBRARY DESTINATION lib
  RUNTIME DESTINATION bin
)

install(
  DIRECTORY include/
  DESTINATION include
)

ament_export_targets(sdformat_urdf-export HAS_LIBRARY_TARGET)
```

sloretz replied:
Interesting! I wonder if more recent CMake is happy because the `LIBRARY DESTINATION` was already set on the target.

I wouldn't want to include the `sdformat_urdf_plugin` target in `sdformat_urdf-export`, so I think it needs to be split into two `install()` rules. I also didn't know about `HAS_LIBRARY_TARGET`.

sloretz replied:
I combined the `sdformat_urdf` target install rules and separated the `sdformat_urdf_plugin` install rule in 86f5f54. Mind trying it on your bionic system?

azeey replied:
Works great!

azeey commented on sdformat_urdf/CMakeLists.txt:
```diff
@@ -0,0 +1,175 @@
+
+ament_export_dependencies(urdfdom_headers)
+ament_export_dependencies(sdformat9)
+
```
I tried using `sdformat_urdf` in another library and it failed saying 
```
Target "test_using_sdformat_urdf" links to target "rcutils::rcutils" but the target was not found.  Perhaps a find_package() call is missing for an IMPORTED target, or an ALIAS target is missing?
```
I think since you have `ament_target_dependencies(sdformat_urdf PUBLIC "rcutils")` [here](https://github.com/ros/sdformat_urdf/pull/1/files#diff-691bc8ec14d67d8f2a9c492b7b2b9e947b0dc037a3fbb6b6df4b5916c80ac47fR37), we should add 
```suggestion
ament_export_dependencies(rcutils)
```
Alternatively, we could make `rcutils` a private dependency, but I'm not sure we can do that using `ament_target_dependencies`.

Here the CMakeLists.txt I used.

```cmake
cmake_minimum_required(VERSION 3.5)
project(test_using_sdformat_urdf)

find_package(sdformat_urdf REQUIRED)

add_executable(test_using_sdformat_urdf src/test_using_sdformat_urdf.cpp)
target_link_libraries(test_using_sdformat_urdf PUBLIC sdformat_urdf::sdformat_urdf)

install(TARGETS test_using_sdformat_urdf RUNTIME DESTINATION bin)
```

sloretz replied:
> Alternatively, we could make rcutils a private dependency, but I'm not sure we can do that using ament_target_dependencies.

I replaced `ament_target_dependencies()` with `target_link_libraries()` in f25ffe2

ament/ament_cmake#292

azeey commented on sdformat_test_files/models/geometry_plane/geometry_plane.sdf:
```diff
@@ -0,0 +1,35 @@
+          </plane>
+        </geometry>
+      </collision>
+      <!-- TODO(sloretz) what's the moment of inertia of a plane? Is it affected by the <normal>? -->
```
I have always used a plane in a static model, so the moment of inertia was irrelevant. I have also thought of it as a half-space, but that's not always how it's implemented. For example, the dartsim plugin in ign-physics uses a large, thin box.

azeey commented on sdformat_test_files/models/joint_revolute2/joint_revolute2.sdf:
```diff
@@ -0,0 +1,80 @@
+          <upper>1.5</upper>
+        </limit>
+      </axis>
+      <axis>
```
Shouldn't this be `<axis2>`?

sloretz replied:
Oops, `axis2` in revolute2 example in 48d9136

azeey commented on sdformat_test_files/models/joint_revolute_two_joints_two_links/joint_revolute_two_joints_two_links.sdf:
```diff
@@ -0,0 +1,84 @@
+      <parent>link_1</parent>
+      <child>link_2</child>
+      <axis>
+        <xyz>1 0 0</xyz>
```
If the `<axis>` here meant to be the same as the `<axis>` in `joint_revolute_1`? The `model.config` file says the two links are being rigidly connected, which suggests that this should an orthogonal joint axis.

sloretz replied:
I think they're rigidly connected because the pose of the joints is different. Neither joint can be rotated because the links are pinned together in two places.

azeey replied:
Ah, makes sense.

azeey commented on sdformat_test_files/models/joint_universal/joint_universal.sdf:
```diff
@@ -0,0 +1,66 @@
+    <joint name="joint_universal" type="universal">
+      <parent>link_1</parent>
+      <child>link_2</child>
+      <pose>0 0 0.1 0 0 0</pose>
```
Does this need `<axis>` and `<axis2>`?

sloretz replied:
The [spec doesn't require them](http://sdformat.org/spec?ver=1.7&elem=joint), but it looks like by default it puts both axis at `0,0,1`, which behaves a little strange in gazebo 11. Added axis elements with perpendicular vectors in 7d0ac7a

azeey commented on sdformat_test_files/models/model_two_models/model_two_models.sdf:
```diff
@@ -0,0 +1,61 @@
+<?xml version='1.0' encoding='utf-8'?>
+<sdf version="1.7">
+  <model name="first_model">
```
The semantics of having multiple models in a model file are not clear to me. I have created an [issue](https://github.com/osrf/sdformat/issues/395) about this.

azeey commented on sdformat_urdf/README.md:
```diff
@@ -0,0 +1,55 @@
+
+### Limitations that may result in Warnings
+
+If any of these constraints are voilated then the library may issue a console warning, but the model is still converted to URDF c++ structures.
```
nit: voilated -> violated

sloretz replied:
voilated -> violated in 2c375de

azeey commented on sdformat_urdf/README.md:
```diff
@@ -0,0 +1,55 @@
+* `<link>` should not have any `<light>` or `<sensor>`
+
+#### Material limitations
+* Only solid color materials are supported - the color of the urdf comes from the `<ambient>` tag of the material
```
Can you elaborate why the `<ambient>` and not the `<diffuse>` tag as chosen? I was under the impression that `<diffuse>` is the "color" of the material

sloretz replied:
IIUC ambient and diffuse are added together to make the final color. Ambient is always visible because it models light that bounced off so many surfaces that the source is unknown. Diffuse is only visible when a light is pointed at it. I went with ambient since it's always visible, but it might not be a great choice if the model was made for an environment with no ambient light, like space.

It could be possible to combine the two in some way, but adding a diffuse component means making an assumption about the color and intensity of the light shining on the object. If the object is modeled for a dark space with bright directional lights then it would make sense to add a lot of the diffuse (which means assume it's illuminated by a bright white light), but If the object was modeled for a bright space with lots of ambient light but dim directional lights then a bright white light might make the final color near-white.

http://gazebosim.org/tutorials?tut=color_model

azeey replied:
Okay. So in that model, I would say diffuse is closer to the "color" of the material because in real life, the color of objects does depend on the light shining on them. In Rviz, the default light is a white light, but [this](https://github.com/ros-visualization/rviz/pull/1146) PR suggests it's possible to add different types of lights. In that case, I would think users want the objects to be affected by the different types of lights available in the environment.

Also from looking at Rviz, it looks like the URDF color is set as both the diffuse and the ambient (with half the intensity). 
https://github.com/ros-visualization/rviz/blob/8bd56fb928b255e095424a5f1fb3ff8a5274602e/src/rviz/robot/robot_link.cpp#L473

sloretz replied:
> I would say diffuse is closer to the "color" of the material because in real life, the color of objects does depend on the light shining on them

`<ambient>` is the same. The color depends on the ambient light color, but ambient light is a [property of the environment rather than a directional light](http://sdformat.org/spec?ver=1.7&elem=scene#scene_ambient). The odd bit about this is the `urdf` color is like the final color one would see when the model is rendered. `sdformat_urdf` using only `<ambient>` as the color is like rendering the model in the scene with `rgba(1 1 1 1)` ambient light and no directional lights. RViz is un-rendering that color to `ambient = color * 0.5` and `diffuse = color` components and re-rendering it to the screen.

Since we need to render to a color, maybe we can use the default lighting setup in Gazebo as the example? The default gazebo empty world has [`rgba(0.4 0.4 0.4 1)` ambient light](http://sdformat.org/spec?ver=1.7&elem=scene#scene_ambient) with a directional light having [diffuse `rgba(0.8 0.8 0.8 1)`](https://github.com/ignitionrobotics/ign-gazebo/blob/da2ddbc14bc487d53e49de5b9fdea3272797698c/examples/worlds/default.sdf#L36) and [specular `rgba(0.2 0.2 0.2 1)`](https://github.com/ignitionrobotics/ign-gazebo/blob/da2ddbc14bc487d53e49de5b9fdea3272797698c/examples/worlds/default.sdf#L37). Specular tends to just be a highlight, so I think that could be ignored. That would leave `urdf color = 0.4 * sdf_ambient + 0.8 * sdf_diffuse`.

sloretz replied:
Uses `0.4 * ambient + 0.8 * diffuse` to render color in c84405d

azeey replied:
Okay, sounds good to me.

azeey commented on sdformat_urdf/package.xml:
```diff
@@ -0,0 +1,42 @@
+  <buildtool_depend>ament_cmake_ros</buildtool_depend>
+  <buildtool_export_depend>ament_cmake_ros</buildtool_export_depend>
+
+  <build_export_depend>urdf</build_export_depend>
```
nit: I'm not super familiar with `package.xml` syntax, but can this just be `<depend>urdf</depend>` since they key is used for `build_depend` and `exec_depend`?

sloretz replied:
[Yup](https://www.ros.org/reps/rep-0140.html#depend-multiple), used `<depend>` in 034edd1

azeey commented on sdformat_urdf/src/sdformat_urdf_plugin.cpp:
```diff
@@ -0,0 +1,75 @@
+      // Since it's an XML document it must have `<sdf>` as the first tag
+      const tinyxml2::XMLElement * root = doc.RootElement();
+      if (std::string("sdf") != root->Name()) {
+        std::cout << "'" << root->Name() << "'\n";
```
nit: Is the `cout` still needed?

sloretz replied:
Oops, removed debug print in abeb693

azeey commented on sdformat_urdf/src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,629 @@
+{
+  auto sdf_dom = std::make_shared<sdf::Root>();
+  errors = sdf_dom->LoadSdfString(data);
+  if (sdf_dom) {
```
It's unlikely that `sdf_dom` will be a `nullptr` since it's allocated in this function. I recommend checking if `errors` is empty instead.

sloretz replied:
Fixed in  ce0812d

azeey commented on sdformat_urdf/src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,629 @@
+urdf::ModelInterfaceSharedPtr
+sdformat_urdf::sdf_to_urdf(const sdf::Root & sdf_dom, sdf::Errors & errors)
+{
+  if (sdf_dom.WorldCount()) {
```
nit: Consider checking against a number since `WorldCount` is not a bool

sloretz replied:
Compare with `0u` in 4681be2

azeey commented on sdformat_urdf/src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,629 @@
+  urdf_link->inertial = std::make_shared<urdf::Inertial>();
+  urdf_link->inertial->mass = sdf_inertia.MassMatrix().Mass();
+  // URDF doesn't have link pose concept, so add SDF link pose to inertial
+  urdf_link->inertial->origin = convert_pose(link_pose + sdf_inertia.Pose());
```
nit: I think using `Pose::operator*` is [preferred](https://github.com/ignitionrobotics/ign-math/issues/60#issuecomment-683193429) instead of `Pose::operator+`.

sloretz replied:
Using `operator*` instead of `operator+` or `operator-` in b06afde

azeey commented on sdformat_urdf/src/sdformat_urdf.cpp:
```diff
@@ -0,0 +1,629 @@
+      // TODO(sloretz) textures
+      // TODO(sloretz) error if any file names we can't resolve are given
+      auto urdf_material = std::make_shared<urdf::Material>();
+      // sdf materials don't have names, so assign it the visual's name and hope that's unique
```
It seems to me that it's actually not that rare for a model to have two links each containing a visual with the same name (eg. `viz`) . How about if we concatenate the name of the link and the visual somehow? Then it should be unique since link names are unique.

sloretz replied:
Concatenate link and visual names for material name in 197bbf8

azeey reviewed (commented):
I have gone through everything except the test cpp files so far. Looks great so far. I appreciate the thorough error checking. Just have some minor comments.

sloretz committed:
check errors instead of valid pointer

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Make urdf a <depend>

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove std::cout dbg print

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
the the -> the

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Compare WorldCount against 0u

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Concatenate link and visual names to make material name

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Compatibility with CMake earlier than 3.14

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use target_link_libraries()

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
One install rule per target

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
voilated -> violated

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use axis2 for second axis

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Universal joint perpendicular axis

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use Pose3d::operator*

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
structures -> DOM objects

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
@azeey I think I addressed all changes requested. Mind having a second look? :bow:

azeey commented on sdformat_urdf/test/include/test_tools.hpp:
```diff
@@ -0,0 +1,87 @@
+    EXPECT_EQ(expected_ign, actual_ign); \
+  } while (false)
+
+#define EXPECT_NAMES(child_ptr_list, ...) \
```
Could you add a line about what this macro does?

sloretz replied:
Documented `EXPECT_NAMES` in e924fe6

azeey commented on sdformat_urdf/test/joint_tests.cpp:
```diff
@@ -0,0 +1,251 @@
+
+  EXPECT_EQ("joint_revolute", joint->name);
+  EXPECT_EQ(urdf::Joint::REVOLUTE, joint->type);
+  EXPECT_DOUBLE_EQ(0.1, joint->axis.x);
```
In libsdformat10 (https://github.com/osrf/sdformat/pull/312), the joint axis will be normalized. Consider changing the test and the corresponding SDFormat file to use unit vector so we don't have to update this when we move to libsdformat10.

sloretz replied:
Compares normalized joint axes in 84d23f5

azeey commented on sdformat_urdf/test/geometry_tests.cpp:
```diff
@@ -0,0 +1,256 @@
+
+  ASSERT_EQ(urdf::Geometry::BOX, link->visual->geometry->type);
+  {
+    urdf::BoxConstSharedPtr box = std::dynamic_pointer_cast<urdf::Box>(link->visual->geometry);
```
Might be good to add an assert for `box`.

sloretz replied:
Added asserts for geometry types not being null after dynamic_pointer_casts in bb65e1f

azeey commented on sdformat_urdf/test/geometry_tests.cpp:
```diff
@@ -0,0 +1,256 @@
+  urdf::LinkConstSharedPtr link = model->getRoot();
+  ASSERT_NE(nullptr, link);
+  ASSERT_NE(nullptr, link->visual);
+  ASSERT_NE(nullptr, link->collision);
```
Could you add an assert for `link->visual->geometry` and `link->collision->geometry` as well? Same for the other tests.

sloretz replied:
Added checks for `->geometry` not being null before using it in 9ae4804

azeey commented on sdformat_urdf/test/joint_tests.cpp:
```diff
@@ -0,0 +1,251 @@
+  ASSERT_TRUE(model);
+  ASSERT_EQ("joint_continuous", model->getName());
+
+  urdf::JointConstSharedPtr joint = model->getJoint("joint_continuous");
```
Could you add an assert here for `joint`?

sloretz replied:
Check joint before using it in d9ca9fb

azeey commented on sdformat_urdf/test/joint_tests.cpp:
```diff
@@ -0,0 +1,251 @@
+  ASSERT_EQ("joint_fixed", model->getName());
+
+  urdf::JointConstSharedPtr joint = model->getJoint("joint_fixed");
+
```
An assert for `joint` here as well and the rest of the tests

sloretz replied:
Check joint before using it in d9ca9fb

azeey commented on sdformat_urdf/test/link_tests.cpp:
```diff
@@ -0,0 +1,107 @@
+  ASSERT_TRUE(model);
+  ASSERT_EQ("link_inertia", model->getName());
+
+  urdf::LinkConstSharedPtr link = model->getLink("link");
```
Could you add an assert for `link` here and in the other tests in this file?

sloretz replied:
Added assert's for link not being null in ec3f81b

azeey reviewed (commented):
I have now gone through all the tests and everything looks great! I think it would be good to add a  test where the inertial, visual and collision are affected by the link's pose. Correct me if I'm wrong, but the tests I have seen had all these elements in the root link so the pose of the link had no impact on the poses of these elements, or the SDFormat link had an identity pose.

sloretz committed:
Document EXPECT_NAMES()

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

gavanderhoorn commented:
Just wanted to take a minute to express my appreciation for the significant amount of work being put into this by @sloretz and the others.

:+1: :robot: :fireworks: 

---

Edit: and as it's off-topic, I'll hide my comment as such.

sloretz committed:
Compare normalized joint axes

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Check casted pointer before using them

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Check geometry pointer isn't null before using it

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Assert joint isn't null before using it

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Assert link isn't null before using it

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Assert link isn't null before using it

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Color = 0.4 * ambient + 0.8 * diffuse

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Fix application of link pose on inertial/visual/collision

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz committed:
Add pose_joint_all test

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>

sloretz commented:
> I think it would be good to add a test where the inertial, visual and collision are affected by the link's pose. Correct me if I'm wrong, but the tests I have seen had all these elements in the root link so the pose of the link had no impact on the poses of these elements, or the SDFormat link had an identity pose.

As in it's missing a test of a link with joint to a link having non-zero pose and non-zero inertial/visual/collision pose? Added `pose_joint_all` model and test in e4664a8b588b7014716610be5e05967c1c1308f5, and fix a bug it discovered in how those poses get calculated in 6582d5221462a9799c4936db45f9b87e5e2f1afd

sloretz commented:
@azeey I think I've addressed all PR comments. Mind having another look?

azeey reviewed (approved):
This looks great! Thanks for iterating with me.
